[TexMap] Limit painting in compositor if damage allows

Scony opened the pull request:
#### 0a22af9ce070d2c72759d50ec8bc11a730452d07
<pre>
[TexMap] Limit painting in compositor if damage allows
<a href="https://bugs.webkit.org/show_bug.cgi?id=286923">https://bugs.webkit.org/show_bug.cgi?id=286923</a>

Reviewed by Carlos Garcia Campos.

This change:
- Stores the Damage in TextureMapper and utilizes it to reduce amount of painting
  done in CoordinatedBackingStore.
- Improves the quality of Damage by changing the way it&apos;s stored and improving the
  internal &quot;merging&quot; step that simplifies the damage when it gets too complex.

By storing the Damage in TextureMapper it&apos;s now possible to leverage the damaging
information during various painting operations. This PR just adds an initial
improvement in the CoordinatedBackingStore that is now able to paint only a
fragment of the tile&apos;s texture when the damage allows.

The Damage has been improved in a way that it doesn&apos;t use expensive Region under
the hood anymore. Instead, it uses just a simple vector of rectangles. The vector,
however, is limited to the predefined size. The rectangles are added just by
adding to the vector until it reaches the predefined size - at this point the
merging algorithm kicks in.

The damage merging algorithm spreads incoming rects among buckets that are nothing
more than a minimum bounding rectangles of all the rects that were put into them.
The decision on which bucket to place the input rectangle in is based on how well
the input rectangle fits into the bucket-corresponding cell of a helper NxM size grid.

* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm-expected.txt: Added.
* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html: Added.
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::region const):
(WebCore::Damage::bounds const):
(WebCore::Damage::rects const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::invalidate):
(WebCore::Damage::add):
(WebCore::Damage::shouldUnite const):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::unite):
(WebCore::Damage::mergeIfNeeded): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawTextureFragment):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
(WebCore::TextureMapper::setDamage):
(WebCore::TextureMapper::damage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelf):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):

Canonical link: <a href="https://commits.webkit.org/291615@main">https://commits.webkit.org/291615@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1dd04459fd9f3e5426f2a2372c7db1fae1d8c87b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93303 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98302 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43829 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13148 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21316 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71303 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28698 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9868 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9559 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2029 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43143 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2045 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100334 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14901 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80324 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20607 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80317 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79639 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19824 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1518 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13478 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20339 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25516 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20026 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -93,8 +93,34 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+            && tile.texture().isOpaque()
+            && !tile.texture().filterOperation()
+        ) {
+            // We define damagedTileRect as a bounding box of all damage rects that intersect tile.rect()
```
```suggestion
            // We define damagedTileRect as a minimum bounding rectangle of all damage rects that intersect tile.rect()
```

Scony commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -505,6 +505,37 @@ void TextureMapper::drawTexture(const BitmapTexture& texture, const FloatRect& t
     drawTexture(texture.id(), texture.colorConvertFlags() | (texture.isOpaque() ? OptionSet<TextureMapperFlags> { } : TextureMapperFlags::ShouldBlend), targetRect, matrix, opacity, allEdgesExposed);
 }
 
+void TextureMapper::drawTextureFragment(const BitmapTexture& sourceTexture, const FloatRect& sourceRect, const FloatRect& targetRect)
```
the name could be better as "fragment" may be misleading (towards fragment shader)

Scony replied:
actually it's hard to find a better name...

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -56,32 +56,84 @@ class Damage {
-    ALWAYS_INLINE IntRect bounds() const { return m_region.bounds(); }
-    ALWAYS_INLINE Rects rects() const { return m_region.rects(); }
-    ALWAYS_INLINE bool isEmpty() const  { return !m_invalid && m_region.isEmpty(); }
+    Region region()
```
```suggestion
    Region region() const
```

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -56,32 +56,84 @@ class Damage {
+
+        m_minimumBoundingRectangle.unite(rect);
+
+        // Artificial NxM grid is used to direct input rectangles into proper buckets (cells)
```
the following feels to me like one particular _damage rectangle merging / coalescing strategy_ out of many different ones we could decide to implement.

I would extract it out into its own compilation unit -- possibly with a recognisable name -- and then potentially write a unit test as it is quite algorithmically. You could verify that given a certain sequence of rectangles, the resulting rectangles are what you expect.

By writing it as a strategy and using the strategy pattern, we could then experiment and benchmark different ones.

Scony replied:
While I agree about unit test (layout test actually), I'm not sure about employing strategy pattern at this point. Wouldn't it be a premature optimization to employ pattern for just a single strategy we have? Please note that we're not even sure atm. if we ever going to have more than 1 strategy.

spenap replied:
We might never have more than one strategy at the same time, but testing new strategies in branches would be simplified. And I am only suggesting extracting this slightly: possibly an interface defining the "adding" method in terms of what it is (coalescing - merging) and this extracted into its own class implementing it.

Or... templates? if you don't want the cost of virtual dispatching?

Scony replied:
I've been thinking about it more and I think we actually should not think of the _merging strategy_ as one of strategies. The truth is, this is just a implementation detail of the data structure (`Damage`).

Before this PR, the `Damage` was acting like a lightweight data structure that was merely a wrapper around `Region`. The only difference was the "valid" API and the merging step that was necessary.

Now, with this PR, `Damage` becomes a data structure on it's own as it stores rects and operates on them in a non-trivial way. In this particular case we also have a merging step, but it's just an effect of the `Damage` internals. Thus it doesn't necessarily have to be present. For example, when working on this PR I've implemented a `RTree` data structure at some point and made `Damage` a simple wrapper of it (just like it was wrapping `Region` before) - in that case there was no "merging" step per se. The amount of the rects that were fetched from the data structure could be directly controlled by doing a particular `RTree` query up to certain level.

The above brings me to other suggestion - IMO we could perhaps employ strategy pattern on the broader scope i.e. we could extract all the implementation details of `Damage` into strategy like e.g. `GridBasedDamage` or so and then make `Damage` be flexible enough so that we could select what "backend" we want to use with particular object. Actually I had such an implementation at some point when experimenting with RTRees and below were the ways to create `Damage` objects:
```
Damage damage(Damage::Purpose::Storage);
Damage damage(Damage::Purpose::Compositor);
// alternatively it could be more implementation-oriented:
Damage damage(Damage::Backend::Vector);
Damage damage(Damage::Backend::RTree);
// or resolution/granularity-oriented
Damage damage(Damage::Resolution::High);
Damage damage(Damage::Resolution::Low);
```

We could ofc. go even further and make `Damage` an interface and prepare some separate implementations but there are pros and cons to that as well.

spenap replied:
Thanks a lot for sharing your thoughts and elaborating so much. Maybe my naming it a strategy was inappropriate, and it's plainly an implementation detail associated to a "backend" as you wrote above.

In any case, what I had in mind was that something similar to that. We could use RTree, Vectors, we could use other criteria... and this implementation as it is in this PR doesn't abstract that away -- it's actually there in the header -- I think that's probably what shocked me the most: to me, it should be behind a better named method or class, as it is a very specific behaviour, which could possibly be supported by design / geometric diagrams explaining how rectangles are combined, and calculating best & worse cases.

It's perfectly possible we don't need to agree on this point at this time, and we can further improve the design / make it more modular in the future.

Scony replied:
I think for now it would be good enough to split this function into smaller ones and keep everything in `Damage.h` to keep things simple and fast.

In the future - we can extract entire storage logic as you said. Technically we can do it as we have this "unified" mode where we unify all the rects into one always - so we already could have 2 "backends" or whatever we'd call it. That, however, should be done in separate PR as it will involve adding even more tests and changing much more than in this PR.

What do you think?

spenap replied:
Sounds good to me!

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -56,32 +56,84 @@ class Damage {
+            const auto tileSize = IntSize(viewportSize / gridSize);
+            const auto rectCenter = rect.center();
+            const auto rectCell = ceiledIntPoint(FloatPoint { float(rectCenter.x()) / tileSize.width(), float(rectCenter.y()) / tileSize.height() });
+            const size_t index = std::clamp(rectCell.x(), 0, int(gridCols - 1)) + std::clamp(rectCell.y(), 0, int(gridRows - 1)) * gridCols;
```
```suggestion
            const size_t index = std::clamp(rectCell.x(), 0, static_cast<int>(gridCols - 1)) + std::clamp(rectCell.y(), 0, static_cast<int>(gridRows - 1)) * gridCols;
```

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -56,32 +56,84 @@ class Damage {
+        constexpr size_t gridRows = 4;
+        constexpr size_t gridCells = gridCols * gridRows;
+        static_assert(gridCols > 0);
+        static_assert(gridRows > 0);
```
Why do we need this if they are fixed values?

Scony replied:
I anticipate those values may be altered by downstream patches in some projects, so just wanted to make sure it's obvious from the code that 0-oes are not accepted - thus a compile-time assertions. If you think those are not necessary, I can remove them - what do you think?

Scony replied:
actually I had to remove those anyway

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -56,32 +56,84 @@ class Damage {
-        m_region.unite(region);
-        mergeIfNeeded();
+        for (const auto& rect : region.rects())
+            add(rect);
```
for each rectangle we invoke the (at first sight) non-trivial `add`. Can this really be `ALWAYS_INLINE`?

Scony replied:
yes, this whole `for` loop can be inlined, but the `add` itself cannot

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -512,6 +512,37 @@ void TextureMapper::drawTexture(const BitmapTexture& texture, const FloatRect& t
+        0
+    ).scale3d(
+        double(sourceRect.width()) / textureSize.width(),
+        double(sourceRect.height()) / textureSize.height(),
```
```suggestion
        static_cast<double>(sourceRect.x()) / textureSize.width(),
        static_cast<double>(sourceRect.y()) / textureSize.height(),
        0
    ).scale3d(
        static_cast<double>(sourceRect.width()) / textureSize.width(),
        static_cast<double>(sourceRect.height()) / textureSize.height(),
```

spenap commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -93,8 +93,34 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
     TransformationMatrix adjustedTransform = transform * TransformationMatrix::rectToRect(layerRect, targetRect);
 
+#if ENABLE(DAMAGE_TRACKING)
+    const auto damageRects = textureMapper.damage().rects();
```
```suggestion
    const auto& damageRects = textureMapper.damage().rects();
```

so we avoid an unnecessary copy

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
```diff
@@ -69,7 +69,31 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
+            && allEdgesExposed == TextureMapper::AllEdgesExposed::No
+            && opacity == 1.0
+            && tile.texture().isOpaque()
+            && !tile.texture().filterOperation()
```
I would move this to a helper function (or lambda) like canUseDamageToDraw() or something similar. Would it be possible in the future to support drawing texture fragment in more cases?

Scony replied:
I'll do it. Yes - in the future I anticipate the fragment drawing will be used in more cases.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapper.h:
```diff
@@ -70,6 +71,7 @@ class TextureMapper {
     void drawNumber(int number, const Color&, const FloatPoint&, const TransformationMatrix&);
 
     WEBCORE_EXPORT void drawTexture(const BitmapTexture&, const FloatRect& target, const TransformationMatrix& modelViewMatrix = TransformationMatrix(), float opacity = 1.0f, AllEdgesExposed = AllEdgesExposed::Yes);
+    void drawTextureFragment(const BitmapTexture& sourceTexture, const FloatRect& sourceRect, const FloatRect& targetRect);
```
Since this is only used by damage I would add `#if ENABLE(DAMAGE_TRACKING)` to make it explicit.

Scony committed:
[TexMap] Limit painting in compositor if damage allows
https://bugs.webkit.org/show_bug.cgi?id=286923

Reviewed by Carlos Garcia Campos.

This change:
- Stores the Damage in TextureMapper and utilizes it to reduce amount of painting
  done in CoordinatedBackingStore.
- Improves the quality of Damage by changing the way it's stored and improving the
  internal "merging" step that simplifies the damage when it gets too complex.

By storing the Damage in TextureMapper it's now possible to leverage the damaging
information during various painting operations. This PR just adds an initial
improvement in the CoordinatedBackingStore that is now able to paint only a
fragment of the tile's texture when the damage allows.

The Damage has been improved in a way that it doesn't use expensive Region under
the hood anymore. Instead, it uses just a simple vector of rectangles. The vector,
however, is limited to the predefined size. The rectangles are added just by
adding to the vector until it reaches the predefined size - at this point the
merging algorithm kicks in.

The damage merging algorithm spreads incoming rects among buckets that are nothing
more than a minimum bounding rectangles of all the rects that were put into them.
The decision on which bucket to place the input rectangle in is based on how well
the input rectangle fits into the bucket-corresponding cell of a helper NxM size grid.

* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm-expected.txt: Added.
* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html: Added.
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::region const):
(WebCore::Damage::bounds const):
(WebCore::Damage::rects const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::invalidate):
(WebCore::Damage::add):
(WebCore::Damage::shouldUnite const):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::unite):
(WebCore::Damage::mergeIfNeeded): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawTextureFragment):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
(WebCore::TextureMapper::setDamage):
(WebCore::TextureMapper::damage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelf):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):

Canonical link: https://commits.webkit.org/291615@main
